# Tutorial


Use handwritting digit recognization data for exhibit how to use retriveStacking

********************************************************************************************************************************************

Load data from scikit-learn dataset:

In [73]:
from sklearn.datasets import load_digits
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import random
random.seed( 3 )


digits =load_digits()
data=pd.DataFrame(digits.data).reset_index()
data['target']=digits.target
predictors=list(data)
predictors.remove('target')


**********************************************************************************************************************************************
Split dataset to train and test (train for training model and test to validate model):

In [74]:
train_X, test_X, train_Y, test_Y = train_test_split(data[predictors], data['target'], test_size=0.333)

**********************************************************************************************************************************************
Load packages; choose several models with default settings; assign models to each layer (3 layer in total,layer 3 is output layer):

In [75]:
from retrieveStack.retrieveEnsemble import ensemble
from retrieveStack.retrieveEnsemble import opt_weight,retrieve
from sklearn.neighbors import KNeighborsClassifier,KNeighborsRegressor
from sklearn.tree import DecisionTreeClassifier,DecisionTreeRegressor
from sklearn.ensemble import RandomForestClassifier, ExtraTreesClassifier,RandomForestRegressor,ExtraTreesRegressor,GradientBoostingClassifier,GradientBoostingRegressor

knc=KNeighborsClassifier()
dtc=DecisionTreeClassifier(random_state=4321)
rfc=RandomForestClassifier(random_state=4321)
etc=ExtraTreesClassifier(random_state=4321)

knr=KNeighborsRegressor()
dtr=DecisionTreeRegressor(random_state=4321)
rfr=RandomForestRegressor(random_state=4321)
etr=ExtraTreesRegressor(random_state=4321)


layer1=[knc,dtc,rfc,etc,knr,dtr,rfr,etr]
layer2=[knc,dtc,rfc,etc,knr,dtr,rfr,etr]
layer3=[knc,dtc,rfc,etc]


**********************************************************************************************************************************************

train first layer, generate input data for next layer and record model information in this layer:

In [76]:
trainX_l1,testX_l1,modelInfo_l1=ensemble(layer_name='layer1',models=layer1,
                                         train_x=train_X,train_y=train_Y,test_x=test_X,
                                         n_classes=10,n_folds=10,
                                         metric_for_classifier='log_loss',
                                         metric_for_regressor='mean_squared_error')

0 KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=5, p=2,
           weights='uniform')
Fold 0 - Fold 1 - Fold 2 - Fold 3 - Fold 4 - Fold 5 - Fold 6 - Fold 7 - Fold 8 - Fold 9 - All folds Done!
the log_loss of layer1 model 0 is 1.86236783862
*******************************************************
1 DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_split=1e-07, min_samples_leaf=1,
            min_samples_split=2, min_weight_fraction_leaf=0.0,
            presort=False, random_state=4321, splitter='best')
Fold 0 - Fold 1 - Fold 2 - Fold 3 - Fold 4 - Fold 5 - Fold 6 - Fold 7 - Fold 8 - Fold 9 - All folds Done!
the log_loss of layer1 model 1 is 5.49982675429
*******************************************************
2 RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=

**********************************************************************************************************************************************
train second layer, generate input data for next layer and record model information in this layer:

In [77]:
trainX_l2,testX_l2,modelInfo_l2=ensemble(layer_name='layer2',models=layer2,
                                         train_x=trainX_l1,train_y=train_Y,test_x=testX_l1,
                                         n_classes=10,n_folds=10,
                                         metric_for_classifier='log_loss',
                                         metric_for_regressor='mean_squared_error')

0 KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=5, p=2,
           weights='uniform')
Fold 0 - Fold 1 - Fold 2 - Fold 3 - Fold 4 - Fold 5 - Fold 6 - Fold 7 - Fold 8 - Fold 9 - All folds Done!
the log_loss of layer2 model 0 is 1.37412274099
*******************************************************
1 DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_split=1e-07, min_samples_leaf=1,
            min_samples_split=2, min_weight_fraction_leaf=0.0,
            presort=False, random_state=4321, splitter='best')
Fold 0 - Fold 1 - Fold 2 - Fold 3 - Fold 4 - Fold 5 - Fold 6 - Fold 7 - Fold 8 - Fold 9 - All folds Done!
the log_loss of layer2 model 1 is 2.46335561796
*******************************************************
2 RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=

**********************************************************************************************************************************************
train third layer, generate input data for next layer and record model information in this layer:

In [78]:
trainX_l3,testX_l3,modelInfo_l3=ensemble(layer_name='layer3',models=layer3,
                                         train_x=trainX_l2,train_y=train_Y,test_x=testX_l2,
                                         n_classes=10,n_folds=10,
                                         metric_for_classifier='log_loss',
                                         metric_for_regressor='mean_squared_error')

0 KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=5, p=2,
           weights='uniform')
Fold 0 - Fold 1 - Fold 2 - Fold 3 - Fold 4 - Fold 5 - Fold 6 - Fold 7 - Fold 8 - Fold 9 - All folds Done!
the log_loss of layer3 model 0 is 1.45967013013
*******************************************************
1 DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_split=1e-07, min_samples_leaf=1,
            min_samples_split=2, min_weight_fraction_leaf=0.0,
            presort=False, random_state=4321, splitter='best')
Fold 0 - Fold 1 - Fold 2 - Fold 3 - Fold 4 - Fold 5 - Fold 6 - Fold 7 - Fold 8 - Fold 9 - All folds Done!
the log_loss of layer3 model 1 is 2.14079537509
*******************************************************
2 RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=

**********************************************************************************************************************************************
Find the good models by comparing score with the output layer (third layer), and combine with output layer's models to weighted average prediction. opt_weight function can give out the cv score after weight averaging and the optimized weight for all the models:

In [79]:
good=retrieve([modelInfo_l1,modelInfo_l2,modelInfo_l3],metric='log_loss',maximize=False)

bestScore,optWeight=opt_weight(modelInfo_list=good,train_y=train_Y,metric='log_loss',method='SLSQP',maximize=False)

The score after weighted averaging is 0.139577532013


**********************************************************************************************************************************************
Output the prediction using optimized weight:

In [80]:
from retrieveStack.retrieveEnsemble import predict_proba,predict
from sklearn.metrics import accuracy_score
pred=predict(good,optWeight,majority_voting=True)
accuracy_score(test_Y,pred)

0.9849749582637729

Compare with the accuracy with single models in first layer:

In [81]:
etc.fit(train_X,train_Y)
accuracy_score(test_Y,etc.predict(test_X))

0.94490818030050083

In [82]:
rfc.fit(train_X,train_Y)
accuracy_score(test_Y,rfc.predict(test_X))

0.95492487479131882

In [83]:
dtc.fit(train_X,train_Y)
accuracy_score(test_Y,dtc.predict(test_X))

0.85642737896494159

In [84]:
knc.fit(train_X,train_Y)
accuracy_score(test_Y,knc.predict(test_X))

0.76627712854757934

In [85]:
knr.fit(train_X,train_Y)
accuracy_score(test_Y,[int(round(e)) for e in knr.predict(test_X)])

0.27378964941569284

In [86]:
dtr.fit(train_X,train_Y)
accuracy_score(test_Y,[int(round(e)) for e in dtr.predict(test_X)])

0.80300500834724542

In [87]:
rfr.fit(train_X,train_Y)
accuracy_score(test_Y,[int(round(e)) for e in rfr.predict(test_X)])

0.60601001669449084

In [88]:
etr.fit(train_X,train_Y)
accuracy_score(test_Y,[int(round(e)) for e in etr.predict(test_X)])

0.62103505843071782

Now we have a look at the result of conventional feedforward stacking (without retrieving back to find good models, only use model from output layer):

In [89]:
bestScore2,optWeight2=opt_weight(modelInfo_list=modelInfo_l3,train_y=train_Y,metric='log_loss',method='SLSQP',maximize=False)

The score after weighted averaging is 0.392044925247


In [90]:
pred=predict(modelInfo_l3,optWeight2,majority_voting=True)
accuracy_score(test_Y,pred)

0.98163606010016691

# Conclusion:

From above it can see, the retrived stacking can improve the accuracy of prediction greatly without tuning any parameter. The accuracy of retrieveStack is 98.5%, while the best single model only gives a 95.3% accuarcy. Furthermore, compared to conventional feedforward stacking without retrieving (accuracy 98.2%), the retrieving process do improve the prediction. 